Source/GameMapEx.cpp

```diff
@@ -117,7 +117,28 @@ int GameMapEx::LoadMap(char *fileName)
 
 void GameMapEx::CopyTileMap(int* tileMap)
 {
+	if (!isInited) {
+		return;
+	}
+
 	int **tileArray = (int**)(mapTileData);
+	OP2Map *p = (OP2Map*)mapObj;
 	
-	memcpy(tileMap, *tileArray, GetMapWidth()*GetMapHeight()*sizeof(int));
+	int padding;
+	int length;
+
+	if (p->clipRect.x1 == -1)
+	{
+		// Around the world map
+		padding = 0;
+		length = p->tileWidth * p->tileHeight;
+	}
+	else
+	{
+		// Standard map
+		padding = p->clipRect.x1 * p->tileHeight;
+		length = p->clipRect.Width() * p->clipRect.Height();
+	}

```

Comment:
I would be hesitant to involve the `clipRect` in this. That's a secondary value, which just happens to coincide with map dimensions. There is no guarantee the `clipRect` will relate to map padding. A map editor could potentially modify `clipRect` independently, should they wish to provide some kind of custom effect.

I would recommend going directly from map dimensions, which are directly tied to padding. I gave some formulas earlier in the thread. They should look something like:
```cpp
std::size_t paddingOffset = mapWidthInTiles >= 512 ? 0 : 32 * mapHeightInTiles;
std::size_t dataSize = mapWidthInTiles * mapHeightInTiles * sizeof(TileData); // sizeof(TileData) == 4
```

Unless perhaps you've encountered something that says otherwise?

----

One additional thought: I don't think `clipRect` relates to the visible portion, but rather the portion over which movement is allowed. Notably, I think the bottom row of the map tends to be excluded from the `clipRect`. In addition, a number of maps mark the bottom row as non passable.

Reply:
That's good to know about clipRect. I may need to double check some C# SDK code as it is making some assumptions based on that being the visible map area.

The issue is that I can't find the visible map size. `tileWidth` returns the double size. I've noticed that `paddingOffsetTileX` returns 32, which may help avoid the magic number you are proposing. Still need the visible size, though. Maybe I can always divide by 2 if `paddingOffsetTileX` is 32?

Reply:
Here's my new proposed code:
```
int mapWidth = p->paddingOffsetTileX == 32 ? p->tileWidth / 2 : p->tileWidth;
memcpy(tileMap, *tileArray + p->paddingOffsetTileX*p->tileHeight, mapWidth*p->tileHeight*sizeof(int));
```

Reply:
Rereading your comment, I'd like to point out that the bottom row is not visible on "on6_01.map". Bottom right coordinates reported from the game's status bar shows 128,127. This is also true on ATW maps: 512,255

So the bottom row of tiles is always not visible on all maps.

Reply:
@TechCor8, would you mind adding a small comment before the function declaration explaining the use case?

I think for the use case of AI calculations, it probably doesn't need the bottom row since it has to be impassable. Maybe there could be some lava or blight on that row, but otherwise, there probably isn't anything to take into account.

Reply:
Huh, you're right. The map editors do show an extra row of tiles at the bottom of maps, which doesn't appear in game. I used to think that was just a difference in coordinate offsets affecting display of coordinates.

I may need to double check some of these data structures against the original code. It's been a while since I've looked at it, so the details may be getting fuzzy. I noticed the data structure in HFL has both `tileWidth`, `tileHeight`, and `logTileWidth`, `logTileHeight`. The methods `GetMapWidth()` and `GetMapHeight()` use the log versions.

Reply:
I was browsing OP2Utility map.h:
```
/**
* \brief	Represents the visible areas of the map.
*
* \note	Maps designated 'around the world' allow for continuous
*			scrolling on the X axis and so will populate X1 with -1
*			and X2 with \c INT_MAX.
*/
Rect clipRect;
```
Comment seems to indicate that the clipRect is the visible area.

Reply:
Did we decide what to do with this?
We can use clipRect or paddingOffsetTileX.

The current code could be slimmed down to:
```
if (p->clipRect.x1 == -1)
    memcpy(tileMap, *tileArray, p->tileWidth*p->tileHeight*sizeof(int));
else
    memcpy(tileMap, *tileArray + p->clipRect.x1*p->tileHeight, p->clipRect.Width()*p->clipRect.Height()*sizeof(int));
```
or use the paddingOffsetTileX and divide for the width:
```
int mapWidth = p->paddingOffsetTileX == 32 ? p->tileWidth / 2 : p->tileWidth;
memcpy(tileMap, *tileArray + p->paddingOffsetTileX*p->tileHeight, mapWidth*p->tileHeight*sizeof(int));
```

Reply:
You know, maybe we should look at merging this, and just opening an issue to revisit any areas we are uncertain about. It seems like the code is far enough along, and sounds like it's basically working.

----

I think I prefer having the single call to `memcpy`, with arguments determined up front.

----

I'm looking at some of the map loading code right now. The tile data loading is checking `Map.paddingOffsetTileX` (offset: `0x38`) to determine if the tile data should be loaded to an offset address. I'll look more into the rest of the details shortly, assuming I don't get called away.

Reply:
Ok, yes, looks like `paddingOffsetTileX` is the value to use to determine padding size, which can replace the hardcoded `32` (if you want). I noticed the value is set when there is padding, but left unset when there is no padding. I assume it is initialized elsewhere to 0.

----

The check to determine if the map is around the world, is when `lgMapTileWidth >= 9` (tileWidth >= 512). If that is not the case (`lgMapTileWidth < 9`), then we have a regular map. For a regular map, the `paddingOffsetTileX` field is set to 32 (hardcoded), and the `lgMapTileWidth` value is incremented by 1. That means if `lgMapTileWidth == 9`, it's unclear if that was the original value, or the incremented value. The adjusted value is stored at offset `0x14`.

When map data is loaded in from the file, it checks `paddingOffsetTileX` to determine the offset to load to. If that value is zero, it simply loads it to the array, without any offset or size adjustment. The size used is `Map.numTiles` (offset: `0x2C`).

If `paddingOffsetTileX` is non-zero, then it calculates `&tileData[paddingOffsetTileX / 32, 0, paddingOffsetTileX % 32]`, and writes the data contiguously to that location. Note that if `paddingOffsetTileX % 32 != 0`, this would behave in a rather unexpected manner, however, due to the internal uses of base 2 log values, a minimum map size of 32 tiles width would prevent that condition from ever being true. For the size, it decrements the `Map.lgTileWidth` field (offset: `0x14`), and uses that to calculate map size. The calculation is: `Map.tileHeight << (Map.lgTileWidth - 1) * sizeof(TileData)`, where `sizeof(TileData) == 4`.

----

Hope that helps.